### ISSUE TITLE
duplicate etcd config.

### ISSUE THREAD

**Issue by PI-Victor** [kubic-project/salt#16]
I think this is why the name of the nodes are messed up on bootstrap. The default install probably makes `--name` default to `hostname`. Then we generate the env var file and restart the service, ~~which makes it pick up the right settings~~ (partially false, the name should reflect the changes). This is also why we get that member directory that causes issues.
I'd suggest using a custom systemd unit file to avoid this in the future, or making sure the service isn't started till the env file is generated.

With:
```
ExecStart=/bin/bash -c "GOMAXPROCS=$(nproc) /usr/sbin/etcd --name=\"${ETCD_NAME}\" --data-dir=\"${ETCD_DATA_DIR}\" --listen-client-urls=\"${ETCD_LISTEN_CLIENT_URLS}\""
```

From the etcd start log
```
...
Mar 14 12:57:21 minion0 etcd[9369]: recognized environment variable ETCD_NAME, but unused: shadowed by corresponding flag 
Mar 14 12:57:21 minion0 etcd[9369]: recognized environment variable ETCD_DATA_DIR, but unused: shadowed by corresponding flag 
Mar 14 12:57:21 minion0 etcd[9369]: recognized environment variable ETCD_LISTEN_CLIENT_URLS, but unused: shadowed by corresponding flag 
```

after removing the flags from systemd service.
```
ExecStart=/bin/bash -c "GOMAXPROCS=$(nproc) /usr/sbin/etcd"
```

```
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://minion0.k8s.local:2379,http://10.17.3.103:2379
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_CA_FILE=/etc/pki/trust/anchors/SUSE_CaaSP_CA.crt
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_CERT_FILE=/etc/pki/minion.crt
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd/
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_DISCOVERY=http://dashboard:2379/v2/keys/_etcd/registry/k8s
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_DISCOVERY_FALLBACK=proxy
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=https://minion0.k8s.local:2380,https://10.17.3.103:2380
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_INITIAL_CLUSTER_STATE=new
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_INITIAL_CLUSTER_TOKEN=k8s
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_KEY_FILE=/etc/pki/minion.key
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_LISTEN_PEER_URLS=https://0.0.0.0:2380
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_NAME=minion0.k8s.local
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_PEER_CA_FILE=/etc/pki/trust/anchors/SUSE_CaaSP_CA.crt
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_PEER_CERT_FILE=/etc/pki/minion.crt
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_PEER_KEY_FILE=/etc/pki/minion.key
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_PEER_TRUSTED_CA_FILE=/etc/pki/trust/anchors/SUSE_CaaSP_CA.crt
Mar 14 12:58:20 minion0 etcd[10075]: recognized and used environment variable ETCD_TRUSTED_CA_FILE=/etc/pki/trust/anchors/SUSE_CaaSP_CA.crt
```

cc: @inercia @ereslibre

**Comment by inercia** [comment]
I think this could be fixed by just removing [this line](https://github.com/kubic-project/salt/blob/master/salt/etcd/etcd.conf.jinja#L7) and let `etcd` use the default name...

**Comment by ereslibre** [comment]
I'm OK with either solution, I don't really understand why we need to provide this arguments `--name=\"${ETCD_NAME}\" --data-dir=\"${ETCD_DATA_DIR}\" --listen-client-urls=\"${ETCD_LISTEN_CLIENT_URLS}\"` to etcd when we are actually sourcing /etc/sysconfig/etcd that will already contain all this information and will be properly used.

As I said, I don't have a strong opinion whether we should remove this arguments from the systemd unit, or remove `ETCD_NAME` from `/etc/sysconfig/etcd` config file.

**Comment by inercia** [comment]
I guess these arguments are provided in command line (in the unit file) because that way they cannot be empty (ie, `--name=""` would produce an error), but I'm not sure...

And regarding the solution, I would prefer to remove this line from [the jinja template](https://github.com/kubic-project/salt/blob/master/salt/etcd/etcd.conf.jinja) just because it would be easier on us.

**Comment by ereslibre** [comment]
AFAIK those arguments will be completely ignored, they are only useful if you want to override the envvars value.

**Comment by ereslibre** [comment]
@PI-Victor How can I reproduce this double registering problems? Because as far as I remember I wasn't hitting this problem when restarting (I always had `minion{n}.k8s.local` registered, never `minion{n}`).

**Comment by PI-Victor** [comment]
so this is how it happens for me.
i bootstrap the cluster with terraform with `MINIONS_SIZE=3 contrib/libvirt/k8s-setup apply`
afterwards i have:

```
vpalade@f151:~> kubectl get node 
NAME                STATUS     AGE
minion0             Ready   37m
minion1             Ready      37m
```

I proceed to rm the `member` dir from `/var/lib/etcd/` to be able to restart the etcd process. The etcd process is running at this state, though. 
After removing the `member` dir, i can restart etcd successfully and then i try to get the nodes again, and i get this.

```
vpalade@f151:~> kubectl get node 
NAME                STATUS     AGE
minion0             NotReady   37m
minion0.k8s.local   Ready      11m
minion1             Ready      37m
```

What i think happens: provision env file -> restart etcd -> etcd becomes member, creates member dir -> discovery service makes it fallback to proxy, creates proxy folder.

The hackish fix you were saying, which i just tested, works like a charm.
in the `etcd.service` unit file:
```
ExecStartPre=/bin/bash -c "if [[ -f /var/lib/etcd/member ]] && [[ -f /var/lib/etcd/proxy ]]; then rm -rf /var/lib/etcd/member; fi"
```

**Comment by inercia** [comment]
So @PI-Victor could we consider this issue as closed by #18 ?

**Comment by PI-Victor** [comment]
@inercia yeah, i'll close this. though the underlying problem still remains. i will look into it and see what's going on in more depth.
